explorer/ast/member.h

```diff
@@ -25,27 +27,114 @@ struct NamedValue {
   Nonnull<const Value*> value;
 };
 
-// A member of a type.
+// A generic member of a type.
+//
+// This is can be a named, positional or other type of member.
+class Member {
+ protected:
+  explicit Member(MemberKind kind) : kind_(kind) {}
+  explicit Member(MemberKind kind, std::string_view name)
+      : name_(name), kind_(kind) {}
+
+ public:
+  virtual ~Member() = default;
+
+  // Prints the Member
+  virtual void Print(llvm::raw_ostream& out) const = 0;
+
+  // Return whether the member's name matches `name`.
+  auto IsNamed(std::string_view name) const -> bool {
+    return name_ && name_.value() == name;
+  }
+
+  // Returns the enumerator corresponding to the most-derived type of this
+  // object.
+  auto kind() const -> MemberKind { return kind_; }
+
+  // The declared type of the member, which might include type variables.
+  virtual auto type() const -> const Value& = 0;
+
+ protected:
+  const std::optional<const std::string_view> name_;
+
+ private:
+  const MemberKind kind_;
+};
+
+// A named member of a type.
 //
 // This is either a declared member of a class, interface, or similar, or a
 // member of a struct with no declaration.
-class Member {
+class NominalMember : public Member {
  public:
-  explicit Member(Nonnull<const Declaration*> declaration);
-  explicit Member(Nonnull<const NamedValue*> struct_member);
+  explicit NominalMember(Nonnull<const Declaration*> declaration);
+  explicit NominalMember(Nonnull<const NamedValue*> struct_member);
+  NominalMember(const NominalMember& other);
+  NominalMember(NominalMember&& other) noexcept;
+  ~NominalMember() override = default;
+
+  // Prints the Member
+  void Print(llvm::raw_ostream& out) const override;
 
+  static auto classof(const Member* member) -> bool {
+    return InheritsFromNominalMember(member->kind());
+  }
+
+  auto type() const -> const Value& override;
   // The name of the member.
   auto name() const -> std::string_view;
-  // The declared type of the member, which might include type variables.
-  auto type() const -> const Value&;
   // A declaration of the member, if any exists.
   auto declaration() const -> std::optional<Nonnull<const Declaration*>>;
 
  private:
-  llvm::PointerUnion<Nonnull<const Declaration*>, Nonnull<const NamedValue*>>
+  const llvm::PointerUnion<Nonnull<const Declaration*>,
+                           Nonnull<const NamedValue*>>
```

Comment:
Agree, and yes, we should be able to. Let me see if that has any adverse side-effects.

Reply:
After looking at the code, and post renaming (now `MemberElement` instead of `NominalElement`), it seems more natural to keep these two alternatives within that class: both are Members (as far as I can tell), and share most traits (except declaration has a `declaration()`). So separating them might make usage more difficult, and otherwise using a common ancestor class seems a bit overkill given the minimal difference.